File: packages/rslint-test-tools/tests/n/rules/no-exports-assign.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-exports-assign', null as any, {
  valid: [
    { code: 'module.exports.foo = 1' },
    { code: 'exports.bar = 2' },
    { code: 'module.exports = {}' },
    { code: 'module.exports = exports = {}' },
    { code: 'exports = module.exports = {}' },
    { code: 'exports = module.exports' },
    { code: 'var exports = {}' },
    { code: 'let exports = {}' },
    { code: 'const exports = {}' },
    // { code: 'function foo(exports) { exports = {}; }' }, // TODO: shadowing support
  ],
  invalid: [
    {
      code: 'exports = {}',
      errors: [
        {
          messageId: 'noExportsAssign',
          line: 1,
          column: 1,
        },
      ],
    },
    {
      code: 'exports = 1',
      errors: [
        {
          messageId: 'noExportsAssign',
          line: 1,
          column: 1,
        },
      ],
    },
  ],
});

File: internal/plugins/n/rules/no_exports_assign/no_exports_assign_test.go
package no_exports_assign

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/typescript/rules/fixtures"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoExportsAssignRule(t *testing.T) {
	rule_tester.RunRuleTester(fixtures.GetRootDir(), "tsconfig.json", t, &NoExportsAssignRule, []rule_tester.ValidTestCase{
		{Code: "module.exports.foo = 1;"},
		{Code: "exports.bar = 2;"},
		{Code: "module.exports = {};"},
		{Code: "module.exports = exports = {};"},
		{Code: "exports = module.exports = {};"},
		{Code: "exports = module.exports;"},
		{Code: "var exports = {};"},   // local variable
		{Code: "let exports = {};"},   // local variable
		{Code: "const exports = {};"}, // local variable
		// {Code: "function foo(exports) { exports = {}; }"}, // shadowed - TODO: implement scope/shadowing check
	}, []rule_tester.InvalidTestCase{
		{
			Code: "exports = {};",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noExportsAssign",
					Line:      1,
					Column:    1,
				},
			},
		},
		{
			Code: "exports = 1;",
			Errors: []rule_tester.InvalidTestCaseError{
				{
					MessageId: "noExportsAssign",
					Line:      1,
					Column:    1,
				},
			},
		},
	})
}
